bpf/aya-bpf/src/maps/ringbuf.rs

```diff
@@ -0,0 +1,152 @@
+use core::{
+    cell::UnsafeCell,
+    mem,
+    mem::MaybeUninit,
+    ops::{Deref, DerefMut},
+};
+
+use crate::{
+    bindings::{bpf_map_def, bpf_map_type::BPF_MAP_TYPE_RINGBUF},
+    helpers::{
+        bpf_ringbuf_discard, bpf_ringbuf_output, bpf_ringbuf_query, bpf_ringbuf_reserve,
+        bpf_ringbuf_submit,
+    },
+    maps::PinningType,
+};
+
+#[repr(transparent)]
+pub struct RingBuf {
+    def: UnsafeCell<bpf_map_def>,
+}
+
+unsafe impl Sync for RingBuf {}
+
+/// A ring buffer entry, returned from [`RingBuf::reserve`].
+/// You must [`submit`] or [`discard`] this entry before this gets dropped.
+///
+/// [`submit`]: RingBufEntry::submit
+/// [`discard`]: RingBufEntry::discard
+#[must_use = "BPF verifier requires ring buffer entries to be either submitted or discarded"]
```

Comment:
The verifier will complain.

For ergonomics, I think we want to commit not discard on Drop

Reply:
What is the best argument for committing by default? Since the verifier will complain, now I'm thinking that we shouldn't do anything. I think we want to teach people to commit manually: the sooner you commit, the faster the ringbuf advances; if you're doing multi-event commits, you need to know that ordering matters etc.

Reply:
Commit order doesn’t matter (much), the ring order is based on reservation order.

Committing by default might be ergonomic because the reserve/commit API is used a lot for efficiency purpose (zero copy), and discard is something rarely wanted in that case.

Reply:
It does matter: if you reserve s0 s1 and s2, if you commit s1 and s2 the buffer won't notify, it will only do so once s0 is committed.

Reply:
I think in regular Rust you would want to borrow the actual RingBuffer in the `RingBufEntry` and have a lifetime on it, so that the entry can't outlive the datastructure. And either commit or discard on drop (probably discard because the value might be unset?).

One more thought: in regular Rust actually both options are not the safest, since one could `std::mem::forget` a RingBufEntry which would then make with questionable what happens with follow-up reservations - that's the lesson of all the scoped-thread and async stuff. A safer option to model the problem is by providing some kind of scoped-based API along:

```rust
fn with_reservation(&mut self, callback: impl FnOnce(&mut MaybeUninit<T>)
```

Seems like the BPF verifier might make some of those things unnecessary, but it it might still be good to follow Rust best-practices.

Reply:
> I think in regular Rust you would want to borrow the actual RingBuffer in the RingBufEntry and have a lifetime on it, so that the entry can't outlive the datastructure.

That's a good point.

> A safer option to model the problem is by providing some kind of scoped-based API along...

IMO, callback based APIs gets annoying very quickly (particularly because they block the ergonomic use of `?` operator etc) and I think it might worth sacrificing a bit of verifier-pass "safety" in favor of ergonomics.

Reply:
> I think in regular Rust you would want to borrow the actual RingBuffer in the `RingBufEntry` and have a lifetime on it, so that the entry can't outlive the datastructure. And either commit or discard on drop (probably discard because the value might be unset?).

See my comment above: I think getting a verifier error is probably better than a silent bug because you forgot to commit and we discarded on drop. Or in other words: the verifier will force you to either commit or discard - I think that's good dx we should keep. What I don't like is that the verifier error will be somewhat hard to correlate with a rust source code line, so if we could improve that, we should do so. But committing or discarding in drop seems like a bad idea.

As for `RingBufEntry` borrowing the map yes that's a good idea, although it should be an immutable borrow, or it won't be possible to reserve multiple entries to do batching.

> One more thought: in regular Rust actually both options are not the safest, since one could `std::mem::forget` a RingBufEntry which would then make with questionable what happens with follow-up reservations - that's the lesson of all the scoped-thread and async stuff. A safer option to model the problem is by providing some kind of scoped-based API along:
> 
> ```rust
> fn with_reservation(&mut self, callback: impl FnOnce(&mut MaybeUninit<T>)
> ```

`&mut` wouldn't work with batching. Also in this case, since the verifier will reject a forgotten entry anyway, I don't think it's worth sacrificing a better API (iterator), to avoid an unsoundness issue that in practice can not happen.

Reply:
Is it actually possible to call `bpf_ringbuf_reserve` multiple times? The use-case to reserve multiple slots before comitting can also be fulfilled by calling it a single time and reserving more than 1 element. The downside of allowing multiple  `.reserve()` calls which do an immutable borrow is that dropping/submitting them will commit the next possible slot, even if the slot actually refers to something else. E.g.

So I think the mutable borrow is maybe still a good hoice, and it can be slightly modified to handle multiple entries. E.g.  `reserve` could be modified to take a count parameter and return a modified `RingBufEntry/RingBufView` that implements `Target = [MaybeUnit<T>]`. Then the caller can write multiple entries. It could also offer a `submit(n: usize`) method which allows to specify how many of the elements would be submitted. Although I'm not sure whether that needs const generic magic to actually work or whether a deref to an array of variable size actually works - probably needs some experimentation. 

Btw: 
> Also in this case, since the verifier will reject a forgotten entry anyway,

I'm curious on what base it actually does that? Does it really check whether reserve/submit calls are properly paired? The case of forgetting something isn't really one of the usual memory issues that those environments are meant to protect against.

Reply:
> Is it actually possible to call bpf_ringbuf_reserve multiple times?

It's perfectly fine to do that. Whether a slot is committed or not is tracked by a flag placed right before the slot's memory.

> dropping/submitting them will commit the next possible slot, even if the slot actually refers to something else

I don't think that happens. `submit` takes the pointer to the slot you're committing.

> I'm curious on what base it actually does that? Does it really check whether reserve/submit calls are properly paired? 

I don't know, honestly. The verifier's message is "unreleased reference", so I'd say it's basically on the same ground as preventing pointers from escaping from a "safe" (non-probe) eBPF scope.